File: app01_java_block/src/main/java/com/gandarych/legalchain/consensus/ConsensusEngine.java
package com.gandarych.legalchain.consensus;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * Holds the registry of {@link ConsensusStrategy} beans and the currently active one.
 * Spring injects every strategy on the classpath, so a new consensus algorithm becomes
 * available API-wide by simply adding a {@code @Component} implementing the interface.
 */
@Service
public class ConsensusEngine {

    private final Map<String, ConsensusStrategy> strategies;
    private volatile ConsensusStrategy active;

    public ConsensusEngine(List<ConsensusStrategy> available,
                           @Value("${legalchain.consensus.default-strategy:POS}") String defaultStrategy) {
        this.strategies = available.stream()
                .collect(Collectors.toUnmodifiableMap(ConsensusStrategy::name, Function.identity()));
        this.active = require(defaultStrategy);
    }

    public ConsensusStrategy active() {
        return active;
    }

    public ConsensusStrategy byName(String name) {
        return require(name);
    }

    public void switchTo(String name) {
        this.active = require(name);
    }

    public List<ConsensusStrategy> available() {
        return strategies.values().stream()
                .sorted(java.util.Comparator.comparing(ConsensusStrategy::name))
                .toList();
    }

    private ConsensusStrategy require(String name) {
        ConsensusStrategy strategy = strategies.get(name.toUpperCase());
        if (strategy == null) {
            throw new IllegalArgumentException("Unknown consensus strategy: " + name
                    + " (available: " + strategies.keySet() + ")");
        }
        return strategy;
    }
}

File: app03_csharp_block/src/LegalChain/Consensus/ConsensusEngine.cs
namespace LegalChain.Consensus;

/// <summary>
/// Holds the registry of <see cref="IConsensusStrategy"/> implementations and the
/// currently active one. The ASP.NET Core container injects every registered strategy,
/// so a new consensus algorithm becomes available API-wide by simply registering one
/// more implementation of the interface.
/// </summary>
public class ConsensusEngine
{
    private readonly Dictionary<string, IConsensusStrategy> _strategies;
    private volatile IConsensusStrategy _active;

    public ConsensusEngine(IEnumerable<IConsensusStrategy> available, string defaultStrategy)
    {
        _strategies = available.ToDictionary(s => s.Name);
        _active = Require(defaultStrategy);
    }

    public IConsensusStrategy Active => _active;

    public IConsensusStrategy ByName(string name) => Require(name);

    public void SwitchTo(string name) => _active = Require(name);

    public IReadOnlyList<IConsensusStrategy> Available()
        => _strategies.Values.OrderBy(s => s.Name, StringComparer.Ordinal).ToList();

    private IConsensusStrategy Require(string name)
    {
        if (!_strategies.TryGetValue(name.ToUpperInvariant(), out var strategy))
        {
            throw new ArgumentException("Unknown consensus strategy: " + name
                + " (available: [" + string.Join(", ", _strategies.Keys) + "])");
        }
        return strategy;
    }
}
